### PR TITLE
Optimize setting initialization with large classpaths

### PR DESCRIPTION
**Pull request by retronym**
PathSetting.hashCode can be expensive when a large classpath
is configured.

This commit avoids calling as much by making `visibleSettings` and
friends list, rather than sets.

It also makes it cheaper by optimizing for the common case of no
append/prepend classpath settings.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/util/ClassPath.scala [hunk 1]
```diff
@@ -128,7 +128,10 @@ object ClassPath {
   def split(path: String): List[String] = (path split pathSeparator).toList.filterNot(_ == "").distinct
 
   /** Join classpath using platform-dependent path separator */
-  def join(paths: String*): String  = paths filterNot (_ == "") mkString pathSeparator
+  def join(paths: String*): String  = paths.toList.filterNot(_ == "") match {
+    case only :: Nil => only
```
Comment (diesalbla) [inline src/compiler/scala/tools/nsc/util/ClassPath.scala]: Would not this case be covered by the second one? Or does this provide some optimisation?

Comment (retronym) [reply src/compiler/scala/tools/nsc/util/ClassPath.scala]: `mkString` is not optimized for the special case of a one-element collection containing a `String`. That would be worthwhile as a separate change on 2.13.x.

**Comment by som-snytt** [comment]
Wow, `userSetSettings` isn't a `Set` any more. Maybe rename it to `userListedListings`. The way `craigslist` has listings.

I noticed `AbsSettings` got a cheap hashCode, could PathSetting also just get a cheapened hashCode?

**Comment by retronym** [comment]
@som-snytt I've added a faster implementation of `Path.hashCode` for good measure.

Thinking about this terrain some more, it seems plain wrong to be using `Settings` of any type as entries in a `HashMap`.  The are mutable and the hashcode as implemented can change. `allSettings` should be a `LinkedHashMap[String, Setting]`...

**Comment by retronym** [comment]
> `allSettings` should be a `LinkedHashMap[String, Setting]`

Alternative proposal in https://github.com/scala/scala/pull/8358

**Pull request closed without merging**